tests/functional/timer/test_transfer_create.py

```diff
@@ -278,24 +328,105 @@ def fake_resolve(dt):
     assert sent_data["start"] == expect_value
 
 
-def test_timer_creation_rejects_data_access_requirement(
-    run_line, ep_for_timer, non_ha_mapped_collection
+@pytest.mark.parametrize("data_access_position", ["source", "destination", "both"])
+@pytest.mark.parametrize(
+    "has_matching_consent", ("neither", "source", "destination", "both")
+)
+def test_timer_creation_supports_data_access_on_source_or_dest(
+    run_line, ep_for_timer, data_access_position, has_matching_consent
 ):
-    # explicit timezone is preserved in the formatted data sent to the service
+    if data_access_position == "source":
+        src = make_non_ha_mapped_collection()
+        dst = ep_for_timer
+    elif data_access_position == "destination":
+        src = ep_for_timer
+        dst = make_non_ha_mapped_collection()
+    elif data_access_position == "both":
+        src = make_non_ha_mapped_collection()
+        dst = make_non_ha_mapped_collection()
+    else:
+        raise NotImplementedError
+
+    userinfo_meta = load_response_set("cli.foo_user_info").metadata
+    identity_id = userinfo_meta["user_id"]
+    # setup the consent tree response
+    if has_matching_consent == "neither":
+        setup_timer_consent_tree_response(identity_id)
+    elif has_matching_consent == "source":
+        setup_timer_consent_tree_response(identity_id, src)
+    elif has_matching_consent == "destination":
+        setup_timer_consent_tree_response(identity_id, dst)
+    elif has_matching_consent == "both":
+        setup_timer_consent_tree_response(identity_id, src, dst)
+    else:
+        raise NotImplementedError
+
+    # determine what we expect to happen
+    # successful submission or prompt?
+    # if there is a prompt, what should it contain?
+    appears_in_consent_prompt = []
+    if has_matching_consent == "both":
+        pass
+    elif has_matching_consent == "neither":
+        if data_access_position in ("source", "both"):
+            appears_in_consent_prompt.append(src)
+        if data_access_position in ("destination", "both"):
+            appears_in_consent_prompt.append(dst)
+    elif has_matching_consent == "source":
+        if data_access_position in ("destination", "both"):
+            appears_in_consent_prompt.append(dst)
+    elif has_matching_consent == "destination":
+        if data_access_position in ("source", "both"):
+            appears_in_consent_prompt.append(src)
+    else:
+        raise NotImplementedError
+    expect_consent_error = bool(appears_in_consent_prompt)
```

Comment:
The giant block above didn't feel great to me, since we're taking a matrix over two parameters and then using that to "build" the test behaviors dynamically, in the test.
If someone sees a good way through this, but while still testing that full matrix, I'm open to ideas.

I did consider building the matrix and these behavioral parameters outside of the test, and then feeding in the whole generated "suite" as a set of test parameters. But it ended up looking basically the same as this, only in module scope rather than inside of test function scope. So I thought it came out worse.

Reply:
Wow, yeah, this is tough.

Your approach seems like the best for this situation, and for one reason: `NotImplementedError`. This clearly addresses the possibility that something was missed. I started to do the math on creating a test function for each possible configuration. While simpler, it would be tougher to guarantee that every possible combination was hit.

Kudos for this approach!

Reply:
Coming back to this, I don't have any new ideas for a better way to write it, and with one person approving of the approach taken, I'm considering this resolved.

It still doesn't "feel" resolved to me, but I think this is the sort of thing which we can live with until/unless it really _needs_ to be rewritten.